### PR TITLE
Add beagleplay support - take 3

### PR DESCRIPTION
**Pull request by delhiryder**
Clean PR, with well documented (and compact) commit history. No conflicts with the master branch.

### CODE REVIEW

File: layers/meta-balena-beaglebone/conf/layer.conf [hunk 1]
```diff
@@ -5,7 +5,7 @@ BBFILES += "${LAYERDIR}/recipes-*/*/*.bb \
 
 BBFILE_COLLECTIONS += "balena-beaglebone"
 BBFILE_PATTERN_balena-beaglebone := "^${LAYERDIR}/"
-BBFILE_PRIORITY_balena-beaglebone = "1337"
+BBFILE_PRIORITY_balena-beaglebone = "1338"
```
Comment (floion) [inline layers/meta-balena-beaglebone/conf/layer.conf:8]: I don't think it's needed to change the priority here. Any reason for the increase?

File: layers/meta-balena-beaglebone/conf/layer.conf [hunk 2]
```diff
@@ -26,9 +26,23 @@ PREFERRED_PROVIDER_u-boot:beagleboard-xm = "u-boot"
 KERNEL_IMAGETYPES:beaglebone-ai64 = "Image"
 KERNEL_CLASSES:beaglebone-ai64 = "kernel-uimage"
 
+PREFERRED_PROVIDER_virtual/kernel:beagleplay = "linux-bb.org"
+PREFERRED_PROVIDER_virtual/bootloader:beagleplay = "u-boot-ti-staging"
```
Comment (floion) [inline layers/meta-balena-beaglebone/conf/layer.conf:30]: This can be removed

File: layers/meta-balena-beaglebone/conf/layer.conf [hunk 3]
```diff
@@ -26,9 +26,23 @@ PREFERRED_PROVIDER_u-boot:beagleboard-xm = "u-boot"
...
+PREFERRED_PROVIDER_u-boot:beagleplay = "u-boot-ti-staging"
```
Comment (floion) [inline layers/meta-balena-beaglebone/conf/layer.conf:31]: This can be removed

File: layers/meta-balena-beaglebone/conf/layer.conf [hunk 4]
```diff
@@ -26,9 +26,23 @@ PREFERRED_PROVIDER_u-boot:beagleboard-xm = "u-boot"
...
+PREFERRED_VERSION_u-boot-tools:beagleplay ?= "2022.01"
+PREFERRED_VERSION_u-boot-tools-native:beagleplay ?= "2022.01"
```
Comment (floion) [inline layers/meta-balena-beaglebone/conf/layer.conf:33]: Why are these changes for u-boot-tools needed?

File: layers/meta-balena-beaglebone/conf/layer.conf [hunk 5]
```diff
@@ -26,9 +26,23 @@ PREFERRED_PROVIDER_u-boot:beagleboard-xm = "u-boot"
...
+
+PREFERRED_PROVIDER_virtual/kernel:beagleplay-k3r5 = "linux-bb.org"
+PREFERRED_PROVIDER_virtual/bootloader:beagleplay-k3r5 = "u-boot-ti-staging"
```
Comment (floion) [inline layers/meta-balena-beaglebone/conf/layer.conf:36]: This can be removed as per the Zulip discussion?

File: layers/meta-balena-beaglebone/conf/layer.conf [hunk 6]
```diff
@@ -26,9 +26,23 @@ PREFERRED_PROVIDER_u-boot:beagleboard-xm = "u-boot"
 KERNEL_IMAGETYPES:beaglebone-ai64 = "Image"
 KERNEL_CLASSES:beaglebone-ai64 = "kernel-uimage"
 
+PREFERRED_PROVIDER_virtual/kernel:beagleplay = "linux-bb.org"
+PREFERRED_PROVIDER_virtual/bootloader:beagleplay = "u-boot-ti-staging"
+PREFERRED_PROVIDER_u-boot:beagleplay = "u-boot-ti-staging"
+PREFERRED_VERSION_u-boot-tools:beagleplay ?= "2022.01"
+PREFERRED_VERSION_u-boot-tools-native:beagleplay ?= "2022.01"
+
+PREFERRED_PROVIDER_virtual/kernel:beagleplay-k3r5 = "linux-bb.org"
```
Comment (floion) [inline layers/meta-balena-beaglebone/conf/layer.conf:35]: Is it necessary to change the kernel also for the k3r5 multiconfig? We should only care about bootloader for that multiconfig machine.

File: layers/meta-balena-beaglebone/recipes-bsp/u-boot/u-boot-ti-staging_%.bbappend [hunk 7]
```diff
@@ -1,10 +1,33 @@
...
+UBOOT_MACHINE = "am335x_evm_config"
+UBOOT_MACHINE:beagleplay = "am62x_evm_a53_defconfig"
+UBOOT_MACHINE:beagleplay-k3r5 = "am62x_evm_r5_defconfig"
+
+# Use BeagleBone's u-boot until u-boot-ti-staging has the BBAI64 support upstreamed.
+BRANCH:beagleplay= "v2023.04-ti-09.00.00.010-BeaglePlay"
+BRANCH:beagleplay-k3r5= "v2023.04-ti-09.00.00.010-BeaglePlay"
+UBOOT_GIT_URI:beagleplay = "git://git.beagleboard.org/beagleboard/u-boot.git"
+UBOOT_GIT_URI:beagleplay-k3r5 = "git://git.beagleboard.org/beagleboard/u-boot.git"
+SRCREV:beagleplay = "4a4f4d88ffb620a3d389900f648abb24008f3ddc"
```
Comment (floion) [inline layers/meta-balena-beaglebone/recipes-bsp/u-boot/u-boot-ti-staging_%.bbappend:13]: Once you make the changes requested in https://github.com/balena-os/balena-beaglebone/pull/666/commits/7b5f976b5176be7c8dfd8c114e971c43bc13aa44 you should also be able to remove the aboves related to u-boot switching

File: layers/meta-balena-beaglebone/recipes-bsp/u-boot/u-boot-ti-staging_%.bbappend [hunk 8]
```diff
@@ -15,4 +38,17 @@ SRC_URI:append:beaglebone-ai64 = " \
...
-}
\ No newline at end of file
+}
+
+# The Beagleplay *also* creates this file for both archs, remove one.
+do_deploy:append:beagleplay () {
+    rm -vf ${DEPLOYDIR}/extra_uEnv.txt
+}
+
+SSTATE_ALLOW_OVERLAP_FILES += "${DEPLOYDIR}/extra_uEnv.txt"
```
Comment (floion) [inline layers/meta-balena-beaglebone/recipes-bsp/u-boot/u-boot-ti-staging_%.bbappend:48]: Why is this needed? I don't think there is another recipe providing this file

File: layers/meta-balena-beaglebone/recipes-kernel/linux-firmware/linux-firmware_%.bbappend [hunk 9]
```diff
@@ -0,0 +1,10 @@
+#SRC_URI = "https://github.com/beagleboard/ti-linux-firmware.git;branch=09.00.00.007"
```
Comment (floion) [inline layers/meta-balena-beaglebone/recipes-kernel/linux-firmware/linux-firmware_%.bbappend:1]: Why was a different version of linux-firmware needed?

File: layers/meta-balena-beaglebone/recipes-kernel/kernel-selftest/kernel-selftest.bbappend [hunk 10]
```diff
@@ -0,0 +1,7 @@
+
+PACKAGECONFIG:append:aarch64 = " watchdog"
+TEST_LIST:append:aarch64 = "\
```
Comment (floion) [inline layers/meta-balena-beaglebone/recipes-kernel/kernel-selftest/kernel-selftest.bbappend:3]: How is this supposed to be used?

File: layers/meta-balena-beaglebone/conf/layer.conf (hunk 1 above)
Comment (delhiryder) [reply layers/meta-balena-beaglebone/conf/layer.conf:8]: I added this as part of trying to disable the `systemd-watchdog` (which is actually set in the `meta-balena` layer, whose BBFILE_PRIORITY I was trying to override). This is no longer required, because I found the *correct* way to do it (using the `watchdog-disable.conf` file.

Comment (delhiryder) [reply layers/meta-balena-beaglebone/conf/layer.conf:8]: Reverting this change (i.e. setting the priority back to 1337)

File: layers/meta-balena-beaglebone/conf/layer.conf (hunk 3 above)
Comment (delhiryder) [reply layers/meta-balena-beaglebone/conf/layer.conf:31]: The `u-boot-ti-staging` provider, along with the preferred versions, seem to be needed (at least until the `Beagleplay` support is added to the `git.ti.com` mainline kernel).

File: layers/meta-balena-beaglebone/conf/layer.conf (hunk 4 above)
Comment (delhiryder) [reply layers/meta-balena-beaglebone/conf/layer.conf:33]: These changes are needed to support the (updated) `u-boot` sources that `u-boot-ti-staging` fetches.

File: layers/meta-balena-beaglebone/conf/layer.conf (hunk 5 above)
Comment (delhiryder) [reply layers/meta-balena-beaglebone/conf/layer.conf:36]: Again, see my response above.

File: layers/meta-balena-beaglebone/conf/layer.conf (hunk 6 above)
Comment (delhiryder) [reply layers/meta-balena-beaglebone/conf/layer.conf:35]: We can certainly try removing these lines. I was simply being consistent in my `PREFERRED_PROVIDER` specification.

File: layers/meta-balena-beaglebone/conf/layer.conf [hunk 11]
```diff
@@ -26,9 +26,23 @@ PREFERRED_PROVIDER_u-boot:beagleboard-xm = "u-boot"
 KERNEL_IMAGETYPES:beaglebone-ai64 = "Image"
 KERNEL_CLASSES:beaglebone-ai64 = "kernel-uimage"
 
+PREFERRED_PROVIDER_virtual/kernel:beagleplay = "linux-bb.org"
```
[the remark this answers is not in the document]
Comment (delhiryder) [reply layers/meta-balena-beaglebone/conf/layer.conf:29]: Sure. Will document this. Although it now looks like we *may* be able to simply use the `linux-beagleboard` provider.

File: layers/meta-balena-beaglebone/recipes-bsp/u-boot/u-boot-ti-staging_%.bbappend (hunk 7 above)
Comment (delhiryder) [reply layers/meta-balena-beaglebone/recipes-bsp/u-boot/u-boot-ti-staging_%.bbappend:13]: See my comments above. We still need this branch (as opposed to the ti mainline).

File: layers/meta-balena-beaglebone/recipes-bsp/u-boot/u-boot-ti-staging_%.bbappend (hunk 8 above)
Comment (delhiryder) [reply layers/meta-balena-beaglebone/recipes-bsp/u-boot/u-boot-ti-staging_%.bbappend:48]: I was seeing errors before I discovered the following trick (from whoever added the BB-AI64 support):

```
# The Beagleplay *also* creates this file for both archs, remove one.
do_deploy:append:beagleplay () {
    rm -vf ${DEPLOYDIR}/extra_uEnv.txt
}
```

I think we can remove this line. Will rebuild and retest.

File: layers/meta-balena-beaglebone/recipes-kernel/linux-firmware/linux-firmware_%.bbappend (hunk 9 above)
Comment (delhiryder) [reply layers/meta-balena-beaglebone/recipes-kernel/linux-firmware/linux-firmware_%.bbappend:1]: This was part of the debugging of the `wlcore` issue (which turned out to be unrelated to the `linux-firmware`). We can probably remove this and retry.

File: layers/meta-balena-beaglebone/recipes-kernel/kernel-selftest/kernel-selftest.bbappend (hunk 10 above)
Comment (delhiryder) [reply layers/meta-balena-beaglebone/recipes-kernel/kernel-selftest/kernel-selftest.bbappend:3]: This builds a `test-watchdog` package in the deploy directory. This helps to quickly identify if the current version of the kernel has a problem with the hardware watchdog (this is still an open issue, btw).

File: layers/meta-balena-beaglebone/conf/layer.conf (hunk 11 above)
[the remark this answers is not in the document]
Comment (delhiryder) [reply layers/meta-balena-beaglebone/conf/layer.conf:29]: @floion - looking at the structure of the `linux-beagleboard_5.10.bb` recipe, we will have to make some additional changes to it (or the *.bbappend file) in order to **not** use the `linux-bb.org` recipes. 

Is it ok if we continue to use the `linux-bb.org` recipes for the beagleplay ? I will document the specific reasons (as requested).

[the remark this answers is not in the document]
Comment (floion) [reply layers/meta-balena-beaglebone/conf/layer.conf:29]: @delhiryder if there are reasons to switch to `linux-bb.org` instead of `linux-beagleboard` I am fine with the change as long as we document why we need to use the new one and why we can't stick to using the old one.

File: layers/meta-balena-beaglebone/recipes-bsp/u-boot/u-boot-ti-staging_%.bbappend (hunk 7 above)
Comment (floion) [reply layers/meta-balena-beaglebone/recipes-bsp/u-boot/u-boot-ti-staging_%.bbappend:13]: @delhiryder what you did here was you switched the git url of the u-boot-ti-staging recipe to point to git://git.beagleboard.org/beagleboard/u-boot.git

But what I was saying is that the u-boot-bb.org recipe in meta-ti is the one that's using this u-boot source: https://git.yoctoproject.org/meta-ti/tree/meta-ti-bsp/recipes-bsp/u-boot/u-boot-bb.org_git.bb?h=kirkstone

**Comment by floion** [comment]
Superseded by https://github.com/balena-os/balena-beaglebone/pull/679
